torch/csrc/jit/codegen/cuda/kernel_ir.cpp

```diff
@@ -50,6 +50,20 @@ Predicate::Predicate(IrBuilderPasskey passkey, Bool* value)
   TORCH_INTERNAL_ASSERT(value != nullptr);
 }
 
+Predicate* Predicate::maybeWithReplacedExpr(
+    const Expr* expr,
+    const Expr* new_expr) const {
+  // Just create an instance with dummy data, real data will be filled below
+  auto result = IrBuilder::create<Predicate>(PredicateType::Vectorize);
+
+  result->ptype_ = ptype_;
+  result->expr_ = (expr_ == expr ? new_expr : expr_);
```

Comment:
I wonder if we really need to have this binding with an expr. 

It seems `expr_` is used when lowering the predicate, which is done by a call to `generateConditional` in lower_predicate.cpp. It seems all the calls are done like `generateConditional(expr->predicate())`, so we could just also pass `generateConditional(expr->predicate(), expr)` and drop the expr variable from `Predicate`.

Reply:
Changing `generateConditional` does not work, because `expr->predicate()->expr()` is not always identical to `expr`. But fortunately, seems that `generateConditional` only access non-predicate fields of `expr`, so we can just put the original expr to it and remove this `maybeWithReplacedExpr`. Note that, after removing `maybeWithReplacedExpr`, when creating a `Predicate`, we must use the original expr, not the modified one. For example:
```C++
// The following code does NOT work
expr = expr->withPredicate(...);
auto predicate = IrBuilder::create<Predicate>(expr);
expr = expr->withWritePredicate(predicate);

// The following code works
auto new_expr = expr->withPredicate(...);
auto predicate = IrBuilder::create<Predicate>(expr); // don't use new_expr here
new_expr = new_expr->withWritePredicate(predicate);
```

Reply:
Hmm, that sounds a little fragile. Did you find why `generateConditional` needs to use the original expr?

Reply:
Yeah, it is fragile. The source is predicate removal: https://github.com/csarofeen/pytorch/blob/bc772661cbdb3b711d8e9854ae9b8b7052e3e4a3/torch/csrc/jit/codegen/cuda/predicate_compute.cpp#L352
If you put an immediate result, you would need to `propagateExprInfo` from the original to immediate result.

Reply:
Oh, I see. Can you please file an issue? I'll try to clean it up later. For now, I think we can move on.

Reply:
Issue filed at https://github.com/csarofeen/pytorch/issues/2033